[Serve][docs] Refresh http-guide (#27779)
ray-project/ray: doc/source/serve/http-guide.md

# HTTP API

This section should help you understand how to:
- send HTTP requests to Serve deployments
- use Ray Serve to integrate with FastAPI
- use customized HTTP Adapters
- choose which feature to use for your use case

(serve-http)=
## Calling Deployments via HTTP

> When you deploy a Serve application, the ingress deployment (the one passed to `serve.run`) will be exposed over HTTP.

Reviewer: link to key concepts

Reviewer 2: Yeah, I would point to the key concept Ingress Deployment link, since the requests are sent to its __call__ method.

```{literalinclude} ../serve/doc_code/http_guide.py
:start-after: __begin_starlette__
:end-before: __end_starlette__
:language: python
```

> Any request to the Serve HTTP server at `/` is routed to the deployment's `__call__` method with a [Starlette Request object](https://www.starlette.io/requests/) as the sole argument. The `__call__` method can return any JSON-serializable object or a [Starlette Response object](https://www.starlette.io/responses/) (e.g., to return a custom status code or custom headers).

Reviewer: ```suggestion
Requests to the Serve HTTP server are routed to the deployment's `__call__` method with a [Starlette Request object](https://www.starlette.io/requests/) as the sole argument. The `__call__` method can return any JSON-serializable object or a [Starlette Response object](https://www.starlette.io/responses/) (e.g., to return a custom status code or custom headers).
```

> Commonly, you just need the API to accept numpy array. You can use Serve's `DAGDriver` to simply the request parsing.

Reviewer: ```suggestion
Often for ML models, you just need the API to accept a `numpy` array. You can use Serve's `DAGDriver` to simply the request parsing.
```

```{literalinclude} ../serve/doc_code/http_guide.py
:start-after: __begin_dagdriver__
:end-before: __end_dagdriver__
:language: python
```

```{note}

> Serve provides a library of HTTP adapters help you avoid boilerplate code. The [later section](serve-http-adapters) dive deeper into how it works.

Reviewer: ```suggestion
Serve provides a library of HTTP adapters help you avoid boilerplate code. The [later section](serve-http-adapters) dives deeper into how these work.
```

Reviewer 2: Serve provides a library of HTTP adapters **to** help you

```{literalinclude} ../serve/doc_code/http_guide.py
:start-after: __begin_fastapi__
:end-before: __end_fastapi__
:language: python
```

Now if you send a request to `/hello`, this will be routed to the `root` method of our deployment. We can also easily leverage FastAPI to define multiple routes with different HTTP methods:

```{literalinclude} ../serve/doc_code/http_guide.py
:start-after: __begin_fastapi_multi_routes__
:end-before: __end_fastapi_multi_routes__
:language: python
```

You can also pass in an existing FastAPI app to a deployment to serve it as-is:

```{literalinclude} ../serve/doc_code/http_guide.py
:start-after: __begin_byo_fastapi__
:end-before: __end_byo_fastapi__
:language: python
```

This is useful for scaling out an existing FastAPI app with no modifications necessary.
Existing middlewares, **automatic OpenAPI documentation generation**, and other advanced FastAPI features should work as-is.

```{note}

> Serve currently do not support WebSocket endpoint. If you have a use case needing it, please [let us know](https://github.com/ray-project/ray/issues/new/choose)!

Reviewer: ```suggestion
Serve currently does not support WebSockets. If you have a use case that requires it, please [let us know](https://github.com/ray-project/ray/issues/new/choose)!
```

```

## Choosing the right HTTP feature

Serve offers a layered approach to expose your model with the right HTTP API.

Considering your use case, you can choose the right level of abstraction:
- If the built-in adapters fit your use case or you can supply your own adapter, use `DAGDriver` with `http_adapter`.
- If you are comfortable working with the raw request object, use `starlette.request.Requests` API.

> - If you want a fully fledge API server with validation and doc generation, use the FastAPI integration.

Reviewer: This is really valuable information, maybe we should put it at the top / in the intro to the section with links to the subsections?
